### PR TITLE
refactor: send cmds to unix socket

### PR DESCRIPTION
**Pull request by tynany**
Interfacing with FRR via UNIX sockets instead of vtysh

https://github.com/tynany/frr_exporter/issues/49

Before:
```
real	0m0.224s
user	0m0.003s
sys	0m0.003s
```

After:
```
real	0m0.038s
user	0m0.008s
sys	0m0.001s
```

### CODE REVIEW

**Comment by tynany** [comment]
@bdrung @dswarbrick here are the changes to move to UNIX sockets

**Comment by dswarbrick** [comment]
Do you really need to start a new Go package (github.com/tynany/gofrrsockets)? I have just packaged [prometheus-frr-exporter](https://ftp-master.debian.org/new/prometheus-frr-exporter_0.2.20-1.html) in Debian, and these changes will require that I also package e.g. github-golang-tynany-gofrrsockets-dev, since the trend is to move away from vendor directories (which are discouraged in Debian packages anyway).

Unless you're sure that other applications are likely to use your gofrrsockets package, it would be more convenient for distros like Debian to package frr_exporter if the gofrrsockets code was contained within the same repo. Perhaps put it in an "internal" directory, to keep it the code at arm's length from the exporter code, in case you change your mind and want to split it later. Packages under the "internal" directory can only be imported from their parent directory's packages.

**Comment by tynany** [comment]
> Do you really need to start a new Go package (github.com/tynany/gofrrsockets)? I have just packaged [prometheus-frr-exporter](https://ftp-master.debian.org/new/prometheus-frr-exporter_0.2.20-1.html) in Debian, and these changes will require that I also package e.g. github-golang-tynany-gofrrsockets-dev, since the trend is to move away from vendor directories (which are discouraged in Debian packages anyway).
> 
> Unless you're sure that other applications are likely to use your gofrrsockets package, it would be more convenient for distros like Debian to package frr_exporter if the gofrrsockets code was contained within the same repo. Perhaps put it in an "internal" directory, to keep the code at arm's length from the exporter code, in case you change your mind and want to split it later. Packages under the "internal" directory can only be imported from their parent directory's packages.

To me, the best practice is to expose something like gofrrsockts as a client library so it's easy to reuse, especially with the lack of documentation around how to interface the FRR's UNIX sockets. I have seen other pojects using `vtysh` and I was going to let those projects know they can use the gofrrsockets client library as a better alternative.

That being said, I am happy to move it to frr_exporter if it makes maintaining the Debian package easier. Can you please help me understand the constraint that will require packaging gofrrsockets as well, even though it is just a client library?

**Comment by tynany** [comment]
@dswarbrick thank you for explaining that. I have moved the package internally in the latest commit.

File: collector/bgp.go [hunk 1]
```diff
@@ -48,7 +48,7 @@ func getBGPDesc() map[string]*prometheus.Desc {
 
 	if *bgpPeerDescs {
 		bgpPeerLabels = append(bgpLabels, "peer", "peer_as", "peer_desc")
-}
+	}
```
Comment (bdrung) [inline collector/bgp.go:51]: I noticed this change which does not belong in this PR. So I created a separate one for the formatting: https://github.com/tynany/frr_exporter/pull/51

File: collector/command.go [hunk 2]
```diff
@@ -11,12 +11,41 @@ import (
...
 	frrVTYSHOptions = kingpin.Flag("frr.vtysh.options", "Additional options passed to vtysh.").Default("").String()
 )
 
+func executeBGPCommand(cmd string) ([]byte, error) {
+	if *vtyshEnable {
+		return execVtyshCommand("-c", cmd)
+	}
+	return socketConn.ExecBGPCmd(cmd)
+}
+func executeOSPFCommand(cmd string) ([]byte, error) {
```
Comment (bdrung) [inline collector/command.go:27]: Missing newline before function.

File: collector/command.go [hunk 3]
```diff
@@ -11,12 +11,41 @@ import (
...
+	if *vtyshEnable {
+		return execVtyshCommand("-c", cmd)
+	}
+	return socketConn.ExecOSPFCmd(cmd)
+}
+func executePIMCommand(cmd string) ([]byte, error) {
```
Comment (bdrung) [inline collector/command.go:33]: Missing newline before function.

File: collector/command.go [hunk 4]
```diff
@@ -11,12 +11,41 @@ import (
...
+	if *vtyshEnable {
+		return execVtyshCommand("-c", cmd)
+	}
+	return socketConn.ExecPIMCmd(cmd)
+}
+func executeVRRPCommand(cmd string) ([]byte, error) {
```
Comment (bdrung) [inline collector/command.go:39]: Missing newline before function.

File: collector/command.go (hunk 2 above)
Comment (dswarbrick) [reply collector/command.go:27]: gofmt should be fixing stuff like that.

Comment (bdrung) [reply collector/command.go:27]: No, neither gofmt nor goimports is that strict on that regard.

File: collector/bgp.go (hunk 1 above)
Comment (bdrung) [reply collector/bgp.go:51]: Thanks for merging #51. Can you rebase this PR on master?

**Comment by bdrung** [comment]
Regarding gofrrsockts: I agree with @dswarbrick. From looking at the code, having it as internal library is the best approach - unless there are other projects that want to use this library. Then having it separately and having it packaged for Debian is worth the effort.

**Comment by bdrung** [comment]
I rebased your PR on master + #53 and tested in our environment. The bgpl2vpn collector fails with following error:
> cannot unmarshal outputs of 'show evpn vni json': invalid character '%' looking for beginning of value

```
$ sudo vtysh -c "show evpn vni json"
{
}
```

**Comment by bdrung** [comment]
Here is my rebase: https://github.com/bdrung/frr_exporter/commits/frr-unix-sockets

Feel free to set your branch to this rebase. Otherwise you have to resolve the conflicts yourself as well. My rebase includes the added newlines.

**Comment by tynany** [comment]
> I rebased your PR on master + #53 and tested in our environment. The bgpl2vpn collector fails with following error:
> 
> > cannot unmarshal outputs of 'show evpn vni json': invalid character '%' looking for beginning of value
> 
> ```
> $ sudo vtysh -c "show evpn vni json"
> {
> }
> ```

Looks like evpn is not configured. Do you think a better error message will help here?

**Comment by bdrung** [comment]
> Looks like evpn is not configured. Do you think a better error message will help here?

When using vtysh instead of the socket, the exporter does not produce an error message. I would expect the evpn collector to silently accept that there is no data.
